useragent/user_agent.go

```diff
@@ -85,13 +85,27 @@ func (u info) String() string {
 
 type data []info
 
+// Validate the key value pair being set in the user agent. Error if invalid.
+func validate(key, value string) error {
+	// DBR versions as set in the `DATABRICKS_RUNTIME_VERSION` environment variable
+	// are not valid semver strings. Eg: 15.1 or client.0. Thus we allow arbitrary
+	// values for the runtime key.
+	if key == "runtime" {
+		return nil
+	}
+	if !isAlphanum(key) {
+		return fmt.Errorf("expected user agent key to be alphanumeric: %s", key)
+	}
+	if !isAlphanum(value) && !isSemVer(value) {
+		return fmt.Errorf("expected user agent value for key %q to be alphanumeric or semver: %s", key, value)
+	}
+	return nil
+}
+
 // With always uses the latest value for a given alphanumeric key.
 // Panics if key or value don't satisfy alphanumeric or semver format.
 func (d data) With(key, value string) data {
```

Comment:
why not a simple `if` statement in the `With` method?

Reply:
It's cleaner this way IMO since "runtime" is the exception case.  We still want to add the key when it's "runtime", meaning the check ends up being something like:
```
 if key != "runtime" && !isAlphanum(value) && !isSemVer(value) {}
```

Reply:
Also better error messages and can unit test this way. With the panic stack trace it's difficult to tell where exactly the validation failed.

Reply:
A refresh of the error messages sounds good, but special casing `runtime` does not.

Reply:
Removed the exception for `runtime`